p2p/listener.go

```diff
@@ -0,0 +1,87 @@
+package p2p
+
+import (
+	"errors"
+	"sync"
+
+	ma "gx/ipfs/QmUxSEGbv2nmYNnfXi7839wwQqTN3kwQeUxe8dTjZWZs7J/go-multiaddr"
+	"gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+)
+
+// Listener listens for connections and proxies them to a target
+type Listener interface {
+	Protocol() protocol.ID
+	ListenAddress() ma.Multiaddr
+	TargetAddress() ma.Multiaddr
+
+	start() error
+
+	// Close closes the listener. Does not affect child streams
+	Close() error
+}
+
+type listenerKey struct {
+	proto  string
+	listen string
+	target string
+}
+
+// ListenerRegistry is a collection of local application proto listeners.
+type ListenerRegistry struct {
+	sync.Mutex
+
+	Listeners map[listenerKey]Listener
+	starting  map[listenerKey]struct{}
+}
+
+// Register registers listenerInfo into this registry and starts it
+func (r *ListenerRegistry) Register(l Listener) error {
+	r.Lock()
+	k := getListenerKey(l)
+
+	if _, ok := r.Listeners[k]; ok {
+		r.Unlock()
+		return errors.New("listener already registered")
+	}
+
+	r.Listeners[k] = l
+	r.starting[k] = struct{}{}
```

Comment:
I think a simple state mutex inside the listener may be simpler. I.e., take it on start before registering, release it after starting, and take it before closing.

Reply:
Alternatively, once we fix the stream handler registration problem, we can probably just:

1. Not use a "listener key". Just put listeners in a list, we never really need to look them up.
2. Fully start listening *before* registering. If there's a clash, listening should fail.

Thoughts?

Reply:
> Fully start listening before registering. If there's a clash, listening should fail.

Well, I guess once https://github.com/libp2p/go-libp2p-host/issues/16 is fixed this would be the right way